examples/n64lib/src/lock.rs

```diff
@@ -0,0 +1,36 @@
+pub struct InterruptLock(u32);
+
+/// Enter a critical section by disabling interrupts. The lock is released when it is dropped.
+/// This lock is re-entrant (may be used recursively).
+pub fn interrupt_lock() -> InterruptLock {
+    let mut sr: u32;
+
+    unsafe {
+        asm!("
+            mfc0    $0,$$12
+            and     $$8,$1
+            mtc0    $$8,$$12"
+            : "=r"(sr)
+            : "i"(!1)
+            : "$8"
+            : "volatile"
+        );
+    }
+
+    InterruptLock(sr)
+}
+
+impl Drop for InterruptLock {
+    fn drop(&mut self) {
+        unsafe {
+            asm!("
+                mtc0    $0,$$12"
+                // todo: Hazards? Two nops after this? ^
```

Comment:
Good question... I'd have to go back over the MIPS docs to answer this, myself. But IIRC the instruction pipeline will be stalled in the presence of hazards, but shouldn't cause undefined behavior?

Reply:
The docs say:

> The hazard related to CP0 does not generate the interlock of the
pipeline. Therefore, control the number of required instructions
by program.

They also say:

> The number of NOP instructions between the instructions related to the CP0
register and TLB, or the number of the instructions independent of the conflict can
be calculated from the following expression, using this table.
> `(Number of destination hazards of instruction A) -
{(Number of source hazards of instruction B) +1}`

And then they present an example:

>As an example, to find the number of instructions required between an MTC0 and
a subsequent MFC0 instruction, this is:
>`(7) - (4 + 1) = 2 instructions`

The table they refer to is on page 621 of this document: http://datasheets.chipdb.org/NEC/Vr-Series/Vr43xx/U10504EJ7V0UMJ1.pdf

So, maybe 2 nops would be enough?